File: ai/src/main/java/org/conductoross/conductor/ai/AIModelProvider.java
package org.conductoross.conductor.ai;

import java.io.File;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

import org.conductoross.conductor.ai.models.LLMWorkerInput;
import org.springframework.core.env.Environment;
import org.springframework.stereotype.Component;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

@Getter
@Component
@Slf4j
public class AIModelProvider {

    private final Map<String, AIModel> providerToLLM = new HashMap<>();

    private String payloadStoreLocation;

    public AIModelProvider(
            List<ModelConfiguration<? extends AIModel>> modelConfigurations, Environment env) {
        String defaultPayloadStoreLocation = System.getProperty("user.home") + "/worker-payload/";

        for (ModelConfiguration<? extends AIModel> modelConfiguration : modelConfigurations) {
            try {
                AIModel llm = modelConfiguration.get();
                payloadStoreLocation =
                        env.getProperty(
                                "conductor.file-storage.parentDir", defaultPayloadStoreLocation);
                boolean result = new File(payloadStoreLocation).mkdirs();
                log.info(
                        "Created directory {} ? {} for storing worker payload data",
                        payloadStoreLocation,
                        result);
                providerToLLM.put(llm.getModelProvider(), llm);
                for (String alias : llm.getProviderAliases()) {
                    providerToLLM.put(alias, llm);
                }
            } catch (Throwable t) {
                log.error("cannot init {} model, reason: {}", modelConfiguration, t.getMessage());
            }
        }
    }

    public AIModel getModel(LLMWorkerInput input) {
        String name = input.getLlmProvider();
        if (name == null) {
            throw new RuntimeException("llmProvider not specified: " + name);
        }
        AIModel model = providerToLLM.get(name);
        if (model == null) {
            throw new RuntimeException("no configuration found for: " + name);
        }
        return model;
    }

    public Consumer<TokenUsageLog> getTokenUsageLogger() {
        return usageLog -> log.info("{}", usageLog);
    }
}

File: ai/src/main/typescript/AIModelProvider.ts
import { AIModel, ModelConfiguration } from './AIModel.js';
import { LLMWorkerInput } from './models/index.js';
import { TokenUsageLog } from './types/index.js';

export class AIModelProvider {
  private readonly providerToLLM = new Map<string, AIModel>();
  readonly payloadStoreLocation: string;

  constructor(modelConfigurations: ModelConfiguration<AIModel>[], payloadStoreDir?: string) {
    this.payloadStoreLocation =
      payloadStoreDir ?? `${process.env['HOME'] ?? '/tmp'}/worker-payload/`;

    for (const config of modelConfigurations) {
      try {
        const llm = config.get();
        this.providerToLLM.set(llm.getModelProvider(), llm);
        for (const alias of llm.getProviderAliases?.() ?? []) {
          this.providerToLLM.set(alias, llm);
        }
      } catch (err) {
        console.error(`Cannot init model: ${(err as Error).message}`);
      }
    }
  }

  getModel(input: LLMWorkerInput): AIModel {
    const name = input.llmProvider;
    if (!name) throw new Error('llmProvider not specified');
    const model = this.providerToLLM.get(name);
    if (!model) throw new Error(`No configuration found for: ${name}`);
    return model;
  }

  getTokenUsageLogger(): (log: TokenUsageLog) => void {
    return (log) => console.info(JSON.stringify(log));
  }
}
